File: android/app/src/main/java/com/timerapp/linkb24/data/DayReport.kt
package com.timerapp.linkb24.data

import java.time.Instant
import java.time.LocalDate
import java.time.format.DateTimeFormatter

private val dayLabelFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("dd.MM.yyyy")

fun formatHm(totalSeconds: Long): String {
    val hours = totalSeconds / 3600
    val minutes = (totalSeconds % 3600) / 60
    return String.format("%02d:%02d", hours, minutes)
}

fun formatDayLabel(dateIso: String): String {
    return LocalDate.parse(dateIso).format(dayLabelFormatter)
}

fun parseIsoDateOrNull(dateIso: String): LocalDate? {
    return runCatching { LocalDate.parse(dateIso.trim()) }.getOrNull()
}

fun buildDayReportMarkdown(
    data: AppDataDto,
    dateIso: String,
    extended: Boolean = false,
    nowMillis: Long = System.currentTimeMillis(),
): String {
    val parsed = parseIsoDateOrNull(dateIso)
        ?: return "Укажите дату в формате ГГГГ-ММ-ДД."
    val normalized = parsed.format(DateTimeFormatter.ISO_LOCAL_DATE)
    val dayLabel = formatDayLabel(normalized)
    val tasks = data.tasks
        .filter { todaySeconds(it, normalized, nowMillis) > 0 }
        .sortedByDescending { todaySeconds(it, normalized, nowMillis) }
    if (tasks.isEmpty()) {
        return "# Отчёт за $dayLabel\n\nЗа $dayLabel время не учтено."
    }
    val totalSeconds = tasks.sumOf { todaySeconds(it, normalized, nowMillis) }
    val lines = mutableListOf(
        "# Отчёт за $dayLabel",
        "",
        "**Итого:** ${formatHm(totalSeconds)}",
        "",
    )
    tasks.forEachIndexed { index, task ->
        if (index > 0) {
            lines.add("")
        }
        lines.addAll(formatTaskSection(task, normalized, extended, nowMillis))
    }
    return lines.joinToString("\n")
}

private fun formatTaskSection(
    task: TaskDto,
    dateIso: String,
    extended: Boolean,
    nowMillis: Long,
): List<String> {
    val seconds = todaySeconds(task, dateIso, nowMillis)
    val lines = mutableListOf("## ${task.title} — ${formatHm(seconds)}")
    val result = task.result.trim()
    if (result.isNotEmpty()) {
        lines += listOf("", "**Результат:** $result")
    }
    if (!extended) {
        return lines
    }
    val description = task.description.trim()
    if (description.isNotEmpty()) {
        lines += listOf("", "### Описание", description)
    }
    val daySessions = task.sessions.filter { sessionLocalDate(it.startedAt) == dateIso }
    if (daySessions.isNotEmpty()) {
        lines += listOf(
            "",
            "### Сессии за день",
            "| Начало | Окончание | Длительность | Комментарий | Передано |",
            "| --- | --- | --- | --- | --- |",
        )
        for (session in daySessions) {
            val comment = session.comment.replace("|", "\\|").replace("\n", " ")
            val transferred = (session.bitrixRecordId ?: "").replace("|", "\\|")
            lines += "| ${formatTaskDateTime(session.startedAt) ?: "—"} | " +
                "${sessionEndedLabel(session)} | " +
                "${formatHm(sessionDurationSeconds(session, nowMillis))} | " +
                "$comment | $transferred |"
        }
    }
    return lines
}

private fun sessionEndedLabel(session: SessionDto): String {
    if (session.endedAt == null) {
        return "идёт"
    }
    return formatTaskDateTime(session.endedAt) ?: "—"
}

private fun sessionDurationSeconds(session: SessionDto, nowMillis: Long): Long {
    val start = parseInstant(session.startedAt) ?: return 0L
    val end = session.endedAt?.let(::parseInstant) ?: Instant.ofEpochMilli(nowMillis)
    return (end.epochSecond - start.epochSecond).coerceAtLeast(0)
}

File: src/timerapp_ag/domain/day_report.py
from __future__ import annotations

from datetime import datetime

from ..models import Session, Task
from .datetime_util import session_local_date
from .formatting import format_day_label, format_hm, format_task_datetime
from .queries import today_seconds, today_total_seconds
from .state import AppState


def _sessions_on_date(task: Task, date_iso: str) -> list[Session]:
    return [
        session
        for session in task.sessions
        if session_local_date(session.started_at) == date_iso
    ]


def _session_ended_label(session: Session) -> str:
    if session.ended_at is None:
        return "идёт"
    return format_task_datetime(session.ended_at)


def _session_transferred_label(session: Session) -> str:
    return session.bitrix_record_id or ""


def _format_task_section(
    task: Task,
    date_iso: str,
    *,
    extended: bool,
    now: datetime | None,
) -> list[str]:
    seconds = today_seconds(task, date_iso, now=now)
    lines = [f"## {task.title} — {format_hm(seconds)}"]
    result = task.result.strip()
    if result:
        lines.append("")
        lines.append(f"**Результат:** {result}")
    if not extended:
        return lines

    description = task.description.strip()
    if description:
        lines.append("")
        lines.append("### Описание")
        lines.append(description)

    day_sessions = _sessions_on_date(task, date_iso)
    if day_sessions:
        lines.append("")
        lines.append("### Сессии за день")
        lines.append("| Начало | Окончание | Длительность | Комментарий | Передано |")
        lines.append("| --- | --- | --- | --- | --- |")
        for session in day_sessions:
            comment = session.comment.replace("|", "\\|").replace("\n", " ")
            transferred = _session_transferred_label(session).replace("|", "\\|")
            lines.append(
                "| {start} | {end} | {duration} | {comment} | {transferred} |".format(
                    start=format_task_datetime(session.started_at),
                    end=_session_ended_label(session),
                    duration=format_hm(session.duration_seconds(now=now)),
                    comment=comment,
                    transferred=transferred,
                )
            )
    return lines


def build_day_report_markdown(
    state: AppState,
    date_iso: str,
    *,
    extended: bool = False,
    now: datetime | None = None,
) -> str:
    now = now or datetime.now()
    day_label = format_day_label(date_iso)
    tasks = [
        task
        for task in state.tasks
        if today_seconds(task, date_iso, now=now) > 0
    ]
    tasks.sort(key=lambda task: today_seconds(task, date_iso, now=now), reverse=True)
    if not tasks:
        return f"# Отчёт за {day_label}\n\nЗа {day_label} время не учтено."

    total_seconds = today_total_seconds(state, date_iso, now=now)
    lines = [
        f"# Отчёт за {day_label}",
        "",
        f"**Итого:** {format_hm(total_seconds)}",
        "",
    ]
    for index, task in enumerate(tasks):
        if index > 0:
            lines.append("")
        lines.extend(
            _format_task_section(task, date_iso, extended=extended, now=now)
        )
    return "\n".join(lines)
